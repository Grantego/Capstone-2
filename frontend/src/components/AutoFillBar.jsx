import React, {useState} from "react"
import { Form, redirect, useSubmit } from "react-router-dom"
import { filterSearch } from "../helpers/helpers"
import {InputGroup, Input, Button, ListGroup, ListGroupItem } from "reactstrap"

export async function action({request}) {
    document.querySelector("#name").value = "";
    let formData = await request.formData()
    const name = formData.get("name")
    return redirect(`/chords/${encodeURIComponent(name)}`)
  }

const AutoFillBar = ({chords}) => {

    const [autofillChords, setAutofillChords] = useState([])
    const [inputValue, setInputValue] = useState("")
    const submit = useSubmit()


    const handleChange = (e) => {
        const {value} = e.target
        setInputValue(value)
        const filtered = filterSearch(value, chords)
        setAutofillChords(filtered)
    }

    const handleSubmit = (e) => {
        e.preventDefault()
        const formData = new FormData(e.target)
        submit(e.target, { replace: true }) // Trigger form submission
        console.log('running')
        setInputValue("") // Reset the input value
        setAutofillChords([])
    }
    return (
        <>
            <Form method="post" onSubmit={handleSubmit}>
                <InputGroup>
                    <Input type="text"
                           placeholder="chord name ex: C, Cm, Cmaj7"
                           aria-label="chord-input"
                           name="name"
                           id="name"
                           value={inputValue}
                           onChange={handleChange}/>
                    <Button type="submit">Search</Button>
                </InputGroup>
            </Form>
            <ListGroup id="autofill-results">
                {autofillChords.length !== chords.length && autofillChords.map(chord => (
                <ListGroupItem action href={`/chords/${encodeURIComponent(chord)}`} aria-label={chord} tag="a" key={chord}>
                    {chord}
                </ListGroupItem>
                ))}
            </ListGroup>
        </>
    )
}

export default AutoFillBar