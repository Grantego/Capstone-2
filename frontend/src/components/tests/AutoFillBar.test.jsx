import React from 'react';
import { fireEvent, render, screen, cleanup } from '@testing-library/react';
import AutoFillBar from '../AutoFillBar';
import { RouterProvider, createMemoryRouter, Routes, Route } from 'react-router-dom';
import { describe, it, beforeEach, expect, afterEach } from 'vitest';


describe("AutoFillBar component", () => {
    const chords = ["E","Emaj7", "D", "F"]
    let router
    
    beforeEach(() => {
        router = createMemoryRouter([
            {
                path: '/',
                element: <AutoFillBar chords={chords}/>,
            },
            {
                path: '/chords/:chord',
                element: <div>Chord Page</div>,
            }
        ], {
            initialEntries: ['/'],
        }); 
    })

    afterEach(() => {
        cleanup()
    })

    it("renders without crashing", () => {
        render(<RouterProvider router={router} />);
    })

    it("matches snapshot", function() {
        const {asFragment} = render(<RouterProvider router={router} />)
        expect(asFragment()).toMatchSnapshot();
      })

    it("displays no chords if input is empty", () => {
        render(<RouterProvider router={router} />);
        expect(screen.queryByLabelText("Emaj7")).toBe(null);
    })

    it("filters out invalid chords, shows valid chords", () => {
        render(<RouterProvider router={router} />);
        fireEvent.change(screen.getByLabelText("chord-input"), {target: {value: 'e'}})
        expect(screen.getByText("E")).toBeInTheDocument()
        expect(screen.getByText("Emaj7")).toBeInTheDocument()
        expect(screen.queryByText("D")).toBe(null)
        expect(screen.queryByText("F")).toBe(null)
    })

    it("clicking search button goes to chords page", () => {
        render(<RouterProvider router={router} />);
        fireEvent.change(screen.getByLabelText("chord-input"), {target: {value: 'e'}})
        fireEvent.click(screen.getByText("E"));

        expect(screen.getByText("Chord Page")).toBeInTheDocument();
    })
})