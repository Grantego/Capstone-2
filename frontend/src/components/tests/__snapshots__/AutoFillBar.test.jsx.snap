// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AutoFillBar component > matches snapshot 1`] = `
<DocumentFragment>
  <form
    action="/"
    method="post"
  >
    <div
      class="input-group"
    >
      <input
        aria-label="chord-input"
        class="form-control"
        id="name"
        name="name"
        placeholder="chord name ex: C, Cm, Cmaj7"
        type="text"
        value=""
      />
      <button
        class="btn btn-secondary"
        type="submit"
      >
        Search
      </button>
    </div>
  </form>
  <ul
    class="list-group"
    id="autofill-results"
  />
</DocumentFragment>
`;
